src/shims/posix/fs.rs

```diff
@@ -1207,32 +1223,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
-    fn linux_readdir64_r(
-        &mut self,
-        dirp_op: &OpTy<'tcx, Tag>,
-        entry_op: &OpTy<'tcx, Tag>,
-        result_op: &OpTy<'tcx, Tag>,
-    ) -> InterpResult<'tcx, i32> {
+    fn linux_readdir64(&mut self, dirp_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
         let this = self.eval_context_mut();
 
-        this.assert_target_os("linux", "readdir64_r");
+        this.assert_target_os("linux", "readdir64");
 
         let dirp = this.read_scalar(dirp_op)?.to_machine_usize(this)?;
 
         // Reject if isolation is enabled.
         if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
-            this.reject_in_isolation("`readdir64_r`", reject_with)?;
-            // Set error code as "EBADF" (bad fd)
-            return this.handle_not_found();
+            this.reject_in_isolation("`readdir`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
```

Comment:
`EOVERFLOW` and `ENOENT` are also possible: https://pubs.opengroup.org/onlinepubs/9699919799/functions/readdir_r.html.  But yeah `EACCES` is not listed.  Considering there's no way the `DIR *` is valid with isolation enabled, `EBADF` might be okay.

In general though POSIX doesn't limit the reported errors to only the documented ones: https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_03

> Implementations may generate error numbers listed here under circumstances other than those described, if and only if all those error conditions can always be treated identically to the error conditions as described in this volume of POSIX.1-2017. Implementations shall not generate a different error number from one required by this volume of POSIX.1-2017 for an error condition described in this volume of POSIX.1-2017, but may generate additional errors unless explicitly disallowed for a particular function.

This is important because I'm sure `readdir()` can generate `EIO` if e.g. you unplug your flash drive while reading :)

Reply:
> EOVERFLOW and ENOENT are also possible

Interesting, my local manpage does not list those.

> In general though POSIX doesn't limit the reported errors to only the documented ones

Oh, I didn't know that. I just remembered in the previous PR the author justified the EBADF here saying that readdir doesn't allow anything else.

However, I also saw code like this in the Rust stdlib:
https://github.com/rust-lang/rust/blob/ac470e95852336172197810a9a4f6d2e8c8b6574/library/std/src/sys/unix/rwlock.rs#L51

Now, that is a slightly different situation (no errno involved, the error code is directly returned).

Anyway I guess I do not have a strong opinion here.

Reply:
Uhh, I am not convinced that POSIX really limits the returned errors to just those ones for `pthread_rwlock_rdlock()`: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_rwlock_rdlock.html.  Unless there's documentation elsewhere that limits them.  I don't think they'd explicitly say "these functions shall not return an error code of [EINTR]" if they meant "these functions shall not return any other error codes".

(This is off topic now but I'm not sure where to move it to, #t-libs on Zulip?)

Reply:
https://github.com/rust-lang/rust/issues/94705